irohad/multi_sig_transactions/state/impl/mst_state.cpp

```diff
@@ -62,34 +82,34 @@ namespace iroha {
     auto state_update = StateUpdateResult{
         std::make_shared<MstState>(MstState::empty(log_, completer_)),
         std::make_shared<MstState>(MstState::empty(log_, completer_))};
-    for (auto &&rhs_tx : rhs.internal_state_) {
+    for (auto &&rhs_tx : rhs.batches_.right | boost::adaptors::map_keys) {
       insertOne(state_update, rhs_tx);
     }
     return state_update;
   }
 
   MstState MstState::operator-(const MstState &rhs) const {
-    return MstState(this->completer_,
-                    set_difference(this->internal_state_, rhs.internal_state_),
-                    log_);
-  }
-
-  bool MstState::operator==(const MstState &rhs) const {
-    return std::all_of(
-        internal_state_.begin(), internal_state_.end(), [&rhs](auto &i) {
-          return rhs.internal_state_.find(i) != rhs.internal_state_.end();
-        });
+    const auto &my_batches = batches_.right | boost::adaptors::map_keys;
+    std::vector<DataType> difference;
+    difference.reserve(boost::size(batches_));
```

Comment:
Maybe call `shrink_to_fit` afterwards then?

Reply:
It would be worth shrinking if we were to keep the vector. But it is discarded with the function scope. Please see the return statement and the new constructor of `MstState`:
https://github.com/hyperledger/iroha/blob/19af4575d9047925e34a29e5effa9afcb4ab377a/irohad/multi_sig_transactions/state/mst_state.hpp#L171-L173
https://github.com/hyperledger/iroha/blob/19af4575d9047925e34a29e5effa9afcb4ab377a/irohad/multi_sig_transactions/state/mst_state.hpp#L90-L91